Installation issue: py-torch@2.0.1

### Steps to reproduce the issue

When installing `py-torch@2.0.1` I ran into https://github.com/pytorch/FBGEMM/issues/1470, in short:
```
  >> 4412    $spack-stage/spack-stage-py-torch-2.0.1-ctn2xlo5obd5dvh7z37skphinw4qjxei/spack-src/third_party/fb
             gemm/include/fbgemm/././UtilsAvx2.h:47:37: error: 'int32_t' in namespace 'std' does not name a type
```
which I could fix by adding this patch
```python
    patch(
        "https://github.com/pytorch/FBGEMM/commit/da01a59556fec9776733bf20aea8fe8fb29cdd3d.patch",
        sha256="473596db44998fdb0c93fed47896bf88c9ade9e202966f03c8d351b853a7eb6e",
        when="@2.0.1 +fbgemm",
        working_dir="third_party/fbgemm",
    )
```

but now, I get new errors for kineto (see below).

```console
$ spack spec -I <spec>
Input spec
--------------------------------
 -   py-torch@2.0.1

Concretized
--------------------------------
 -   py-torch@2.0.1%gcc@13.2.1~caffe2~cuda~debug+distributed+fbgemm+gloo+kineto~metal+mkldnn+mpi+nnpack+numa+numpy+onnx_ml+openmp+qnnpack~rocm+tensorpipe~test+valgrind+xnnpack build_system=python_pip patches=473596d,65060b5,6730561,beb3fb5 arch=linux-fedora39-skylake
[+]      ^cmake@3.27.7%gcc@13.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-fedora39-skylake
[+]          ^curl@8.4.0%gcc@13.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-fedora39-skylake
[+]              ^nghttp2@1.57.0%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^ncurses@6.4%gcc@13.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-fedora39-skylake
[+]          ^zlib-ng@2.1.4%gcc@13.2.1+compat+opt build_system=autotools arch=linux-fedora39-skylake
[+]      ^eigen@3.4.0%gcc@13.2.1~ipo build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-fedora39-skylake
[+]          ^gmake@4.4.1%gcc@13.2.1~guile build_system=generic arch=linux-fedora39-skylake
[+]      ^fp16@2020-05-14%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^fxdiv@2020-04-17%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^gloo@2023-01-17%gcc@13.2.1~cuda~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^intel-oneapi-mkl@2023.2.0%gcc@13.2.1~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-fedora39-skylake
[+]          ^intel-tbb@2021.9.0%gcc@13.2.1~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make patches=91755c6 arch=linux-fedora39-skylake
[+]      ^ninja@1.11.1%gcc@13.2.1+re2c build_system=generic arch=linux-fedora39-skylake
[+]          ^re2c@2.2%gcc@13.2.1 build_system=generic arch=linux-fedora39-skylake
[+]      ^numactl@2.0.14%gcc@13.2.1 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-fedora39-skylake
[+]          ^autoconf@2.69%gcc@13.2.1 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-fedora39-skylake
[+]          ^automake@1.16.5%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^libtool@2.4.7%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^m4@1.4.19%gcc@13.2.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-fedora39-skylake
[+]              ^diffutils@3.9%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]              ^libsigsegv@2.14%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]      ^openmpi@4.1.6%gcc@13.2.1~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~internal-pmix~java~legacylaunchers~lustre~memchecker~openshmem~orterunprefix+romio+rsh~singularity+static+vt+wrapper-rpath build_system=autotools fabrics=none schedulers=none arch=linux-fedora39-skylake
[+]          ^hwloc@2.9.1%gcc@13.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-fedora39-skylake
[+]              ^libpciaccess@0.17%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]                  ^util-macros@1.19.3%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]              ^libxml2@2.10.3%gcc@13.2.1+pic~python+shared build_system=autotools arch=linux-fedora39-skylake
[+]          ^openssh@9.5p1%gcc@13.2.1+gssapi build_system=autotools arch=linux-fedora39-skylake
[+]              ^krb5@1.20.1%gcc@13.2.1+shared build_system=autotools arch=linux-fedora39-skylake
[+]                  ^bison@3.8.2%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]                  ^findutils@4.9.0%gcc@13.2.1 build_system=autotools patches=440b954 arch=linux-fedora39-skylake
[+]              ^libedit@3.1-20210216%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^perl@5.38.0%gcc@13.2.1+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-fedora39-skylake
[+]              ^berkeley-db@18.1.40%gcc@13.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-fedora39-skylake
[+]          ^pkgconf@1.9.5%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^pmix@5.0.1%gcc@13.2.1~docs+pmi_backwards_compatibility~python~restful build_system=autotools arch=linux-fedora39-skylake
[+]              ^libevent@2.1.12%gcc@13.2.1+openssl build_system=autotools arch=linux-fedora39-skylake
[+]      ^protobuf@3.24.3%gcc@13.2.1~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-fedora39-skylake
[+]          ^abseil-cpp@20230125.3%gcc@13.2.1~ipo+shared build_system=cmake build_type=Release cxxstd=14 generator=make arch=linux-fedora39-skylake
[+]      ^psimd@2020-05-17%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^pthreadpool@2021-04-13%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^py-astunparse@1.6.3%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-six@1.16.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-filelock@3.12.4%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-hatch-vcs@0.3.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-hatchling@1.18.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]              ^py-editables@0.3%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]              ^py-packaging@23.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]              ^py-pathspec@0.11.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]              ^py-pluggy@1.0.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]              ^py-trove-classifiers@2023.8.7%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]                  ^py-calver@2022.6.26%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-jinja2@3.1.2%gcc@13.2.1~i18n build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-markupsafe@2.1.3%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-networkx@2.7.1%gcc@13.2.1+default~extra build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-numpy@1.26.2%gcc@13.2.1 build_system=python_pip patches=873745d arch=linux-fedora39-skylake
[+]          ^py-cython@0.29.36%gcc@13.2.1 build_system=python_pip patches=c4369ad arch=linux-fedora39-skylake
[+]          ^py-pyproject-metadata@0.7.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-pip@23.1.2%gcc@13.2.1 build_system=generic arch=linux-fedora39-skylake
[+]      ^py-protobuf@3.20.3%gcc@13.2.1~cpp build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-pybind11@2.10.1%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^py-pyyaml@6.0%gcc@13.2.1+libyaml build_system=python_pip arch=linux-fedora39-skylake
[+]          ^libyaml@0.2.5%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]      ^py-requests@2.31.0%gcc@13.2.1~socks build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-certifi@2023.7.22%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-charset-normalizer@3.3.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-idna@3.4%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-urllib3@2.0.6%gcc@13.2.1~brotli~secure~socks build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-setuptools@68.0.0%gcc@13.2.1 build_system=generic arch=linux-fedora39-skylake
[+]      ^py-sympy@1.11.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-mpmath@1.2.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-tqdm@4.66.1%gcc@13.2.1~notebook~telegram build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-setuptools-scm@7.1.0%gcc@13.2.1+toml build_system=python_pip arch=linux-fedora39-skylake
[+]              ^git@2.42.0%gcc@13.2.1+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=linux-fedora39-skylake
[+]                  ^libidn2@2.3.4%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]                      ^libunistring@1.1%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]                  ^pcre2@10.42%gcc@13.2.1~jit+multibyte build_system=autotools arch=linux-fedora39-skylake
[+]              ^py-tomli@2.0.1%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-typing-extensions@4.8.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]          ^py-flit-core@3.9.0%gcc@13.2.1 build_system=python_pip arch=linux-fedora39-skylake
[+]      ^py-wheel@0.37.1%gcc@13.2.1 build_system=generic arch=linux-fedora39-skylake
[+]      ^python@3.11.6%gcc@13.2.1+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-fedora39-skylake
[+]          ^bzip2@1.0.8%gcc@13.2.1~debug~pic+shared build_system=generic arch=linux-fedora39-skylake
[+]          ^expat@2.5.0%gcc@13.2.1+libbsd build_system=autotools arch=linux-fedora39-skylake
[+]              ^libbsd@0.11.7%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]                  ^libmd@1.0.4%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^gdbm@1.23%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^gettext@0.22.3%gcc@13.2.1+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-fedora39-skylake
[+]              ^libiconv@1.17%gcc@13.2.1 build_system=autotools libs=shared,static arch=linux-fedora39-skylake
[+]              ^tar@1.34%gcc@13.2.1 build_system=autotools zip=pigz arch=linux-fedora39-skylake
[+]                  ^pigz@2.7%gcc@13.2.1 build_system=makefile arch=linux-fedora39-skylake
[+]                  ^zstd@1.5.5%gcc@13.2.1+programs build_system=makefile compression=none libs=shared,static arch=linux-fedora39-skylake
[+]          ^libffi@3.4.4%gcc@13.2.1 build_system=autotools patches=070b1f3 arch=linux-fedora39-skylake
[+]          ^libxcrypt@4.4.35%gcc@13.2.1~obsolete_api build_system=autotools patches=4885da3 arch=linux-fedora39-skylake
[+]          ^openssl@3.1.3%gcc@13.2.1~docs+shared build_system=generic certs=mozilla arch=linux-fedora39-skylake
[+]              ^ca-certificates-mozilla@2023-05-30%gcc@13.2.1 build_system=generic arch=linux-fedora39-skylake
[+]          ^readline@8.2%gcc@13.2.1 build_system=autotools patches=bbf97f1 arch=linux-fedora39-skylake
[+]          ^sqlite@3.43.2%gcc@13.2.1+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-fedora39-skylake
[+]          ^util-linux-uuid@2.38.1%gcc@13.2.1 build_system=autotools arch=linux-fedora39-skylake
[+]          ^xz@5.4.1%gcc@13.2.1~pic build_system=autotools libs=shared,static arch=linux-fedora39-skylake
[+]      ^sleef@3.5.1_2020-12-22%gcc@13.2.1~ipo build_system=cmake build_type=Release generator=ninja arch=linux-fedora39-skylake
[+]      ^valgrind@3.20.0%gcc@13.2.1+boost+mpi+only64bit~ubsan build_system=autotools libs=shared,static arch=linux-fedora39-skylake
[+]          ^boost@1.83.0%gcc@13.2.1+atomic+chrono~clanglibcpp~container~context~contract~coroutine~date_time~debug+exception~fiber~filesystem~graph~graph_parallel~icu~iostreams~json~locale~log~math~mpi+multithreaded~nowide~numpy~pic~program_options~python~random~regex~serialization+shared~signals~singlethreaded~stacktrace+system~taggedlayout~test+thread~timer~type_erasure~versionedlayout~wave build_system=generic cxxstd=98 patches=a440f96 visibility=hidden arch=linux-fedora39-skylake
```


### Error message

<details><summary>Error message</summary>

<pre>
  [4273/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/DaemonConfigLoader.cpp.o
  In file included from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:12,
                   from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/IpcFabricConfigClient.h:21,
                   from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/DaemonConfigLoader.cpp:16:
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'bool dynolog::ipcfabric::EndPoint<kMaxNumFds>::tryPeekMsg(TCtxt&) [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:155:33:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:174:5: warning: 'throw' will always call 'terminate' [-Wterminate]
    174 |     throw std::runtime_error(std::strerror(errno));
        |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'const char* dynolog::ipcfabric::EndPoint<kMaxNumFds>::getName(const TCtxt&) const [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:170:43:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:185:9: warning: 'throw' will always call 'terminate' [-Wterminate]
    185 |         throw std::invalid_argument(
        |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    186 |             "Unexpected socket name: " + std::string(ctxt.msg_name.sun_path) +
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    187 |             ". Expected to start with " + std::string(socket_dir));
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:192:9: warning: 'throw' will always call 'terminate' [-Wterminate]
    192 |         throw std::invalid_argument(
        |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    193 |             "Expected abstract socket, got " +
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    194 |             std::string(ctxt.msg_name.sun_path));
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'bool dynolog::ipcfabric::EndPoint<kMaxNumFds>::tryRcvMsg(TCtxt&) [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:178:34:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:160:5: warning: 'throw' will always call 'terminate' [-Wterminate]
    160 |     throw std::runtime_error(std::strerror(errno));
        |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  [4274/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/ConfigLoader.cpp.o
  [4275/5588] Building CXX object third_party/ideep/mkl-dnn/src/backend/dnnl/CMakeFiles/dnnl_graph_backend_dnnl.dir/patterns/matmul_fusion.cpp.o
  [4276/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_api.dir/src/ThreadUtil.cpp.o
  [4277/5588] Building CXX object third_party/fmt/CMakeFiles/fmt.dir/src/format.cc.o
  [4278/5588] Linking CXX static library lib/libfmt.a
  [4279/5588] Building CXX object third_party/ideep/mkl-dnn/src/interface/CMakeFiles/dnnl_graph_common.dir/op_schema.cpp.o
  [4280/5588] Building CXX object c10/CMakeFiles/c10.dir/core/AutogradState.cpp.o
  [4281/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/AbstractConfig.cpp.o
  [4282/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/init.cpp.o
  [4283/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_csv.cpp.o
  FAILED: third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_csv.cpp.o
  $spack/lib/spack/env/gcc/g++ -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_EXTERNAL_MZCRC -D_FILE_OFFSET_BITS=64 -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/onnx -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/build/third_party/onnx -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/foxi -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/build/third_party/foxi -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/include -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/fmt/include -I$spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric -I/extras/CUPTI/include -I/include -I/include/roctracer -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/build/third_party/gloo -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/cmake/../third_party/gloo -isystem $spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/protobuf-3.24.3-7ax6j5pt3yepb4yxr5vtwx64k5pbvxv7/include -isystem $spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/intel-oneapi-mkl-2023.2.0-3v5bjqcyqhco3wcojdzzhxm7ynu2o6hz/mkl/latest/include -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/gemmlowp -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/neon2sse -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/XNNPACK/include -isystem $spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/numactl-2.0.14-ercygklomdz5gluhnepcquictckdf6m4/include -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/ittapi/include -isystem $spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/eigen-3.4.0-m2g6ungxix5hecbxjffhackapanlziv2/include/eigen3 -isystem $spack/opt/spack/linux-fedora39-skylake/gcc-13.2.1/openmpi-4.1.6-r7jfc625krddeosyzeoxpgt3qbrpjswu/include -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/ideep/mkl-dnn/third_party/oneDNN/include -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/ideep/include -isystem $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/ideep/mkl-dnn/include -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -O3 -DNDEBUG -DNDEBUG -std=c++14 -fPIC -DCAFFE2_USE_GLOO -DTH_HAVE_THREAD -DKINETO_NAMESPACE=libkineto -DFMT_HEADER_ONLY -DENABLE_IPC_FABRIC -std=c++14 -MD -MT third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_csv.cpp.o -MF third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_csv.cpp.o.d -o third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_csv.cpp.o -c $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/output_csv.cpp
  In file included from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/output_csv.h:10,
                   from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/output_csv.cpp:9:
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/SampleListener.h:96:13: error: variable or field 'init' declared void
     96 | inline void SampleValue::init(uint64_t v) {
        |             ^~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/SampleListener.h:96:31: error: 'uint64_t' was not declared in this scope
     96 | inline void SampleValue::init(uint64_t v) {
        |                               ^~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/SampleListener.h:16:1: note: 'uint64_t' is defined in header '<cstdint>'; did you forget to '#include <cstdint>'?
     15 | #include <vector>
    +++ |+#include <cstdint>
     16 |
  [4284/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/ActivityType.cpp.o
  [4285/5588] Building CXX object c10/CMakeFiles/c10.dir/core/Allocator.cpp.o
  [4286/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/IpcFabricConfigClient.cpp.o
  In file included from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:12,
                   from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/IpcFabricConfigClient.h:21,
                   from $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/src/IpcFabricConfigClient.cpp:11:
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'bool dynolog::ipcfabric::EndPoint<kMaxNumFds>::tryPeekMsg(TCtxt&) [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:155:33:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:174:5: warning: 'throw' will always call 'terminate' [-Wterminate]
    174 |     throw std::runtime_error(std::strerror(errno));
        |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'const char* dynolog::ipcfabric::EndPoint<kMaxNumFds>::getName(const TCtxt&) const [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:170:43:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:185:9: warning: 'throw' will always call 'terminate' [-Wterminate]
    185 |         throw std::invalid_argument(
        |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    186 |             "Unexpected socket name: " + std::string(ctxt.msg_name.sun_path) +
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    187 |             ". Expected to start with " + std::string(socket_dir));
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:192:9: warning: 'throw' will always call 'terminate' [-Wterminate]
    192 |         throw std::invalid_argument(
        |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
    193 |             "Expected abstract socket, got " +
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    194 |             std::string(ctxt.msg_name.sun_path));
        |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h: In instantiation of 'bool dynolog::ipcfabric::EndPoint<kMaxNumFds>::tryRcvMsg(TCtxt&) [with long unsigned int kMaxNumFds = 0; TCtxt = dynolog::ipcfabric::EndPointCtxt<0>]':
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/FabricManager.h:178:34:   required from here
  $spack-stage/spack-stage-py-torch-2.0.1-kq2c75zaxxxttm6q5lobtzqyepqh4sar/spack-src/third_party/kineto/libkineto/third_party/dynolog/dynolog/src/ipcfabric/Endpoint.h:160:5: warning: 'throw' will always call 'terminate' [-Wterminate]
    160 |     throw std::runtime_error(std::strerror(errno));
        |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  [4287/5588] Building CXX object c10/CMakeFiles/c10.dir/core/CopyBytes.cpp.o
  [4288/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/ActivityProfilerProxy.cpp.o
  [4289/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/ILoggerObserver.cpp.o
  [4290/5588] Building CXX object c10/CMakeFiles/c10.dir/core/DefaultDtype.cpp.o
  [4291/5588] Building CXX object c10/CMakeFiles/c10.dir/core/CPUAllocator.cpp.o
  [4292/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/GenericTraceActivity.cpp.o
  [4293/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/ActivityProfilerController.cpp.o
  [4294/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/CuptiActivityProfiler.cpp.o
  [4295/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/Logger.cpp.o
  [4296/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/Config.cpp.o
  [4297/5588] Building CXX object third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_json.cpp.o
  [4298/5588] Building CXX object third_party/ideep/mkl-dnn/src/backend/dnnl/CMakeFiles/dnnl_graph_backend_dnnl.dir/patterns/single_op_pattern.cpp.o
  ninja: build stopped: subcommand failed.
</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (3cf7f7b8000d788d596865de2c2cb8bc52e651b9)
* **Python:** 3.12.0
* **Platform:** linux-fedora39-skylake
* **Concretizer:** clingo

### Additional information

@adamjstewart

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

Hello has this been addressed?

> ==> Error: ProcessError: Command exited with status 1:
>     '/software/spack_environments/default/00/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-13.1.0/python-3.10.13-yt5k3c4siymqh5ppb3lmti5pdsem56fp/bin/python3.10' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/software/spack_environments/scratch/fg12/play1/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-13.1.0/py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn' '.'
> 
> 22 errors found in build log:
>      4355      [3547/5587] Building CXX object confu-deps/XNNPACK/CMakeFiles/convolution-test-helpers.dir/test/convolution-test-helpers.cc.o
>      4356      [3548/5587] Building C object confu-deps/XNNPACK/CMakeFiles/XNNPACK.dir/src/init.c.o
>      4357      [3549/5587] Linking CXX static library lib/libXNNPACK.a
>      4358      [3550/5587] Building CXX object third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmBfloat16Convert.cc.o
>      4359      [3551/5587] Building CXX object third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFloat16Convert.cc.o
>      4360      [3552/5587] Building CXX object third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFPCommon.cc.o
>   >> 4361    FAILED: third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFPCommon.cc.o
>      4362      /software/spack_environments/scratch/fg12/play1/lib/spack/env/gcc/g++ -DFBGEMM_STATIC -I/tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/c
>              puinfo/include -I/tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/third_party/asmjit/src -I/tmp/fg12/spack-stage/spack-stage-py-torch
>              -2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include -I/tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm -isys
>              tem /software/spack_environments/scratch/fg12/play1/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-13.1.0/protobuf-3.24.3-cxnck3jv5iibo52dofnpgfo6bunxekxk/include -isystem /software/spack_environments
>              /scratch/fg12/play1/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-13.1.0/openblas-0.3.24-g4ego7ccgiqdrtabebgh4rgoraecbkik/include -isystem /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchs
>              c6gvrf7cdujeibitgfahn/spack-src/third_party/gemmlowp -isystem /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/neon2sse -isystem /tmp/fg12/s
>              pack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/XNNPACK/include -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPO
>              OL -fopenmp -Wall -Wextra -Werror -Wno-deprecated-declarations -O3 -DNDEBUG -std=c++14 -fPIC -fvisibility=hidden -MD -MT third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFPCommon.cc.o -
>              MF third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFPCommon.cc.o.d -o third_party/fbgemm/CMakeFiles/fbgemm_generic.dir/src/FbgemmFPCommon.cc.o -c /tmp/fg12/spack-stage/spack-stage-py-t
>              orch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/src/FbgemmFPCommon.cc
>      4363      In file included from /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include/fbgemm/./Utils.h:10,
>      4364                       from /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include/fbgemm/SimdUtils.h:9,
>      4365                       from /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include/fbgemm/FbgemmPackMatrixB.h:17,
>      4366                       from /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include/fbgemm/FbgemmFPCommon.h:9,
>      4367                       from /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/src/FbgemmFPCommon.cc:8:
>   >> 4368    /tmp/fg12/spack-stage/spack-stage-py-torch-2.0.1-sbf6bnmkchsc6gvrf7cdujeibitgfahn/spack-src/third_party/fbgemm/include/fbgemm/././UtilsAvx2.h:47:37: error: 'int32_t' in namespace 'std' does not n
>              ame a type
>      4369         47 | template <typename BIAS_TYPE = std::int32_t>
>      4370            |                                     ^~~~~~~
>

--- a/var/spack/repos/builtin/packages/pthreadpool/package.py
+++ b/var/spack/repos/builtin/packages/pthreadpool/package.py
@@ -15,7 +15,8 @@ class Pthreadpool(CMakePackage):
     license("BSD-2-Clause")
 
     version("master", branch="master")
-    version("2021-04-13", commit="a134dd5d4cee80cce15db81a72e7f929d71dd413")  # py-torch@1.9
+    version("2023-08-29", commit="4fe0e1e183925bf8cfa6aae24237e724a96479b8")  # py-torch@2.2
+    version("2021-04-13", commit="a134dd5d4cee80cce15db81a72e7f929d71dd413")  # py-torch@1.9:2.1
     version("2020-10-05", commit="fa75e65a58a5c70c09c30d17a1fe1c1dff1093ae")  # py-torch@1.8
     version("2020-06-15", commit="029c88620802e1361ccf41d1970bd5b07fd6b7bb")  # py-torch@1.6:1.7
     version("2019-10-29", commit="d465747660ecf9ebbaddf8c3db37e4a13d0c9103")  # py-torch@1.5
@@ -35,8 +36,8 @@ class Pthreadpool(CMakePackage):
     )
     resource(
         name="googletest",
-        url="https://github.com/google/googletest/archive/release-1.10.0.zip",
-        sha256="94c634d499558a76fa649edb13721dce6e98fb1e7018dfaeba3cd7a083945e91",
+        url="https://github.com/google/googletest/archive/release-1.12.0.zip",
+        sha256="ce7366fe57eb49928311189cb0e40e0a8bf3d3682fca89af30d884c25e983786",
         destination="deps",
         placement="googletest",
     )

--- a/var/spack/repos/builtin/packages/py-torch/package.py
+++ b/var/spack/repos/builtin/packages/py-torch/package.py
@@ -23,9 +23,10 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     # core libraries to ensure that the package was successfully installed.
     import_modules = ["torch", "torch.autograd", "torch.nn", "torch.utils"]
 
-    license("Intel")
+    license("BSD-3-Clause")
 
     version("main", branch="main")
+    version("2.2.0", tag="v2.2.0", commit="8ac9b20d4b090c213799e81acf48a55ea8d437d6")
     version("2.1.2", tag="v2.1.2", commit="a8e7c98cb95ff97bb30a728c6b2a1ce6bff946eb")
     version("2.1.1", tag="v2.1.1", commit="4c55dc50355d5e923642c59ad2a23d6ad54711e7")
     version("2.1.0", tag="v2.1.0", commit="7bcf7da3a268b435777fe87c7794c382f444e86d")
@@ -129,9 +130,9 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     )
 
     # Required dependencies
-    # See python_min_version in setup.py
-    # Upper bounds come from wheel availability on PyPI
-    depends_on("python@3.8:3.11", when="@2:", type=("build", "link", "run"))
+    # Based on PyPI wheel availability
+    depends_on("python@3.8:3.12", when="@2.2:", type=("build", "link", "run"))
+    depends_on("python@3.8:3.11", when="@2.0:2.1", type=("build", "link", "run"))
     depends_on("python@:3.10", when="@1.11:1", type=("build", "link", "run"))
     depends_on("python@:3.9", when="@1.7.1:1.10", type=("build", "link", "run"))
     depends_on("python@:3.8", when="@1.4:1.7.0", type=("build", "link", "run"))
@@ -155,6 +156,7 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
 
     # setup.py
     depends_on("py-filelock", when="@2:", type=("build", "run"))
+    depends_on("py-typing-extensions@4.8:", when="@2.2:", type=("build", "run"))
     depends_on("py-typing-extensions@3.6.2.1:", when="@1.7:", type=("build", "run"))
     depends_on("py-sympy", when="@2:", type=("build", "run"))
     depends_on("py-networkx", when="@2:", type=("build", "run"))
@@ -190,7 +192,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     # depends_on("sleef@3.5.1_2020-12-22", when="@1.8:")
     # depends_on("sleef@3.4.0_2019-07-30", when="@1.6:1.7")
     depends_on("fp16@2020-05-14", when="@1.6:")
-    depends_on("pthreadpool@2021-04-13", when="@1.9:")
+    depends_on("pthreadpool@2023-08-29", when="@2.2:")
+    depends_on("pthreadpool@2021-04-13", when="@1.9:2.1")
     depends_on("pthreadpool@2020-10-05", when="@1.8")
     depends_on("pthreadpool@2020-06-15", when="@1.6:1.7")
     depends_on("psimd@2020-05-17", when="@1.6:")
@@ -198,13 +201,15 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("benchmark", when="@1.6:+test")
 
     # Optional dependencies
-    # https://github.com/pytorch/pytorch#prerequisites
+    # cmake/public/cuda.cmake
     depends_on("cuda@11:", when="@2:+cuda", type=("build", "link", "run"))
     depends_on("cuda@10.2:", when="@1.11:1+cuda", type=("build", "link", "run"))
     # https://discuss.pytorch.org/t/compiling-1-10-1-from-source-with-gcc-11-and-cuda-11-5/140971
     depends_on("cuda@10.2:11.4", when="@1.10+cuda", type=("build", "link", "run"))
     depends_on("cuda@9.2:11.4", when="@1.6:1.9+cuda", type=("build", "link", "run"))
     depends_on("cuda@9:11.4", when="@:1.5+cuda", type=("build", "link", "run"))
+    # https://github.com/pytorch/pytorch#prerequisites
+    depends_on("cudnn@8.5:", when="@2.3:+cudnn")
     depends_on("cudnn@7:", when="@1.6:+cudnn")
     depends_on("cudnn@7", when="@:1.5+cudnn")
     depends_on("magma+cuda", when="+magma+cuda")
@@ -244,7 +249,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("gloo@2020-03-17", when="@1.6+gloo")
     depends_on("gloo+cuda", when="@1.6:+gloo+cuda")
     # https://github.com/pytorch/pytorch/issues/60331
-    # depends_on("onnx@1.14.1", when="@2.1:+onnx_ml")
+    # depends_on("onnx@1.15.0", when="@2.2:+onnx_ml")
+    # depends_on("onnx@1.14.1", when="@2.1+onnx_ml")
     # depends_on("onnx@1.13.1", when="@2.0+onnx_ml")
     # depends_on("onnx@1.12.0", when="@1.13:1+onnx_ml")
     # depends_on("onnx@1.11.0", when="@1.12+onnx_ml")
@@ -258,6 +264,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("py-hypothesis", type="test")
     depends_on("py-six", type="test")
     depends_on("py-psutil", type="test")
+
+    conflicts("%gcc@:9.3", when="@2.2:", msg="C++17 support required")
 
     # https://github.com/pytorch/pytorch/issues/90448
     patch(
@@ -424,6 +432,13 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
         "https://patch-diff.githubusercontent.com/raw/pytorch/pytorch/pull/110063.patch?full_index=1",
         sha256="23fb4009f7337051fc5303927ff977186a5af960245e7212895406477d8b2f66",
         when="@:2.1",
+    )
+
+    patch(
+        "https://github.com/pytorch/FBGEMM/commit/da01a59556fec9776733bf20aea8fe8fb29cdd3d.patch?full_index=1",
+        sha256="97d8bd43f8cd8bb203dab3480d609c08499224acaca9915f2bdeb23c62350fb1",
+        when="@2.0.1 +fbgemm",
+        working_dir="third_party/fbgemm",
     )
 
     @when("@1.5.0:")

--- a/var/spack/repos/builtin/packages/py-torchaudio/package.py
+++ b/var/spack/repos/builtin/packages/py-torchaudio/package.py
@@ -17,6 +17,7 @@ class PyTorchaudio(PythonPackage):
     license("BSD-2-Clause")
 
     version("main", branch="main")
+    version("2.2.0", tag="v2.2.0", commit="08901ade5d17d3e3cf6fc039cbd601cbd2853686")
     version("2.1.2", tag="v2.1.2", commit="c4c1957d24b423200fd83591d46066135979a5a8")
     version("2.1.1", tag="v2.1.1", commit="db624844f5c95bb7618fe5a5f532bf9b68efeb45")
     version("2.1.0", tag="v2.1.0", commit="6ea1133706801ec6e81bb29142da2e21a8583a0a")
@@ -42,8 +43,9 @@ class PyTorchaudio(PythonPackage):
     version("0.5.0", tag="v0.5.0", commit="09494ea545738538f9db2dceeffe10d421060ee5")
     version("0.4.0", tag="v0.4.0", commit="8afed303af3de41f3586007079c0534543c8f663")
 
-    # https://pytorch.org/audio/main/installation.html#dependencies
-    depends_on("python@3.8:3.11", when="@2:", type=("build", "link", "run"))
+    # Based on PyPI wheel availability
+    depends_on("python@3.8:3.12", when="@2.2:", type=("build", "link", "run"))
+    depends_on("python@3.8:3.11", when="@2.0:2.1", type=("build", "link", "run"))
     depends_on("python@:3.10", when="@0.12:0", type=("build", "link", "run"))
     depends_on("python@:3.9", when="@0.7.2:0.11", type=("build", "link", "run"))
     depends_on("python@:3.8", when="@:0.7.0", type=("build", "link", "run"))
@@ -60,6 +62,7 @@ class PyTorchaudio(PythonPackage):
     depends_on("sox")
 
     depends_on("py-torch@main", when="@main", type=("build", "link", "run"))
+    depends_on("py-torch@2.2.0", when="@2.2.0", type=("build", "link", "run"))
     depends_on("py-torch@2.1.2", when="@2.1.2", type=("build", "link", "run"))
     depends_on("py-torch@2.1.1", when="@2.1.1", type=("build", "link", "run"))
     depends_on("py-torch@2.1.0", when="@2.1.0", type=("build", "link", "run"))

--- a/var/spack/repos/builtin/packages/py-torchtext/package.py
+++ b/var/spack/repos/builtin/packages/py-torchtext/package.py
@@ -19,6 +19,7 @@ class PyTorchtext(PythonPackage):
     license("BSD-3-Clause")
 
     version("main", branch="main")
+    version("0.17.0", tag="v0.17.0", commit="400da5c61bab4abaaeaeca91744ca031ad9b2edf")
     version("0.16.2", tag="v0.16.2", commit="299b90e908c1b492139a4cf9da3912660e79a06b")
     version("0.16.1", tag="v0.16.1", commit="66671007c84e07386da3c04e5ca403b8a417c8e5")
     version("0.16.0", tag="v0.16.0", commit="4e255c95c76b1ccde4f6650391c0bc30650d6dbe")
@@ -38,7 +39,7 @@ class PyTorchtext(PythonPackage):
     version("0.6.0", tag="0.6.0", commit="3a54c7f52584f201c17ca7489b52b812152612dc")
     version("0.5.0", tag="0.5.0", commit="0169cde2f1d446ae886ef0be07e9a673585ed256")
 
-    # https://github.com/pytorch/text#installation
+    # Based on PyPI wheel availability
     depends_on("python@3.8:3.11", when="@0.15:", type=("build", "link", "run"))
     depends_on("python@:3.10", when="@0.13:0.14", type=("build", "link", "run"))
     depends_on("python@:3.9", when="@0.8.1:0.12", type=("build", "link", "run"))
@@ -62,6 +63,7 @@ class PyTorchtext(PythonPackage):
 
     # https://github.com/pytorch/text#installation
     depends_on("py-torch@main", when="@main", type=("build", "link", "run"))
+    depends_on("py-torch@2.2.0", when="@0.17.0", type=("build", "link", "run"))
     depends_on("py-torch@2.1.2", when="@0.16.2", type=("build", "link", "run"))
     depends_on("py-torch@2.1.1", when="@0.16.1", type=("build", "link", "run"))
     depends_on("py-torch@2.1.0", when="@0.16.0", type=("build", "link", "run"))

--- a/var/spack/repos/builtin/packages/py-torchvision/package.py
+++ b/var/spack/repos/builtin/packages/py-torchvision/package.py
@@ -19,6 +19,7 @@ class PyTorchvision(PythonPackage):
     license("BSD-3-Clause")
 
     version("main", branch="main")
+    version("0.17.0", sha256="55e395d5c7d9bf7658c82ac633cac2224aa168e1bfe8bb5b2b2a296c792a3500")
     version("0.16.2", sha256="8c1f2951e98d8ada6e5a468f179af4be9f56d2ebc3ab057af873da61669806d7")
     version("0.16.1", sha256="d31fe52e4540750c8d372b0f38f1bfa81d8261193f2c2c06577332831d203c50")
     version("0.16.0", sha256="79b30b082237e3ead21e74587cedf4a4d832f977cf7dfeccfb65f67988b12ceb")
@@ -53,14 +54,16 @@ class PyTorchvision(PythonPackage):
     variant("ffmpeg", default=False, description=desc.format("FFMPEG"))
     variant("video_codec", default=False, description=desc.format("video_codec"))
 
-    # https://github.com/pytorch/vision#installation
-    depends_on("python@3.8:3.11", when="@0.15:", type=("build", "link", "run"))
+    # Based on PyPI wheel availability
+    depends_on("python@3.8:3.12", when="@0.17:", type=("build", "link", "run"))
+    depends_on("python@3.8:3.11", when="@0.15:0.16", type=("build", "link", "run"))
     depends_on("python@:3.10", when="@0.12:0.14", type=("build", "link", "run"))
     depends_on("python@:3.9", when="@0.8.2:0.11", type=("build", "link", "run"))
     depends_on("python@:3.8", when="@0.5:0.8.1", type=("build", "link", "run"))
 
     # https://github.com/pytorch/vision#installation
     depends_on("py-torch@main", when="@main", type=("build", "link", "run"))
+    depends_on("py-torch@2.2.0", when="@0.17.0", type=("build", "link", "run"))
     depends_on("py-torch@2.1.2", when="@0.16.2", type=("build", "link", "run"))
     depends_on("py-torch@2.1.1", when="@0.16.1", type=("build", "link", "run"))
     depends_on("py-torch@2.1.0", when="@0.16.0", type=("build", "link", "run"))

